contracts/proposal/cwd-proposal-delegate/src/contract.rs

```diff
@@ -0,0 +1,238 @@
+};
+use crate::state::{Config, Delegation, CONFIG, DELEGATIONS, DELEGATION_COUNT, EXECUTE_CTX};
+
+/*
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:cwd-proposal-delegate";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+*/
+
+const DEFAULT_POLICY_IRREVOCABLE: bool = false;
+const DEFAULT_POLICY_PRESERVE_ON_FAILURE: bool = false;
+
+const REPLY_ID_EXECUTE_PROPOSAL_HOOK: u64 = 0;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    let admin = deps.api.addr_validate(&msg.admin)?;
+
+    CONFIG.save(deps.storage, &Config { admin })?;
+
+    Ok(Response::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Delegate {
+            delegate,
+            msgs,
+            expiration,
+            policy_irrevocable,
+            policy_preserve_on_failure,
+        } => {
+            let policy_irrevocable = policy_irrevocable.unwrap_or(DEFAULT_POLICY_IRREVOCABLE);
+            let policy_preserve_on_failure =
+                policy_preserve_on_failure.unwrap_or(DEFAULT_POLICY_PRESERVE_ON_FAILURE);
+            let delegate = deps.api.addr_validate(&delegate)?;
+            execute_delegate(
+                deps,
+                env,
+                info,
+                Delegation {
+                    delegate,
+                    msgs,
+                    expiration,
+                    policy_irrevocable,
+                    policy_preserve_on_failure,
+                },
+            )
+        }
+        ExecuteMsg::RemoveDelegation { delegation_id } => {
+            execute_remove_delegation(deps, env, info, delegation_id)
+        }
+        ExecuteMsg::Execute { delegation_id } => execute_execute(deps, env, info, delegation_id),
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn reply(deps: DepsMut, _env: Env, msg: Reply) -> Result<Response, ContractError> {
+    match msg.id {
+        REPLY_ID_EXECUTE_PROPOSAL_HOOK => {
+            let id = EXECUTE_CTX.load(deps.storage)?;
+            let Delegation {
+                policy_preserve_on_failure,
+                ..
+            } = DELEGATIONS.load(deps.storage, id)?;
+
+            if let SubMsgResult::Err(_) = msg.result {
+                // && with `let` expression above gives unstable Rust error
+                if policy_preserve_on_failure {
+                    return Ok(Response::default()
+                        .add_attribute("execute_failed_but_preserved", id.to_string()));
+                }
+            }
+            // Delete delegation in both success and error case
+            DELEGATIONS.remove(deps.storage, id);
+            Ok(Response::default())
```

Comment:
nit: nice to have this in an `else` block because it makes it clear from reading that we'll never "fall out" of the if statement above and into this.